### PR TITLE
binutils: Only embed `zlib` if listed as a build dependency

### PR DESCRIPTION
**Pull request by geimer**
Addresses https://github.com/easybuilders/easybuild-easyblocks/issues/1350 in conjunction with https://github.com/easybuilders/easybuild-easyconfigs/pull/8340

This does *not* address the issue that `libz.a` is included in `libbfd.a` when specified as a build dependency. At the moment, I don't believe that this is fixable at all without rewriting parts of the Makefile(s).

### CODE REVIEW

File: easybuild/easyblocks/b/binutils.py [hunk 1]
```diff
@@ -175,8 +180,9 @@ def sanity_check_step(self):
                 os.path.join('include', 'libiberty.h'),
             ])
 
-        # if zlib is listed as a dependency, it should get linked in statically
-        if get_software_root('zlib'):
+        # if zlib is listed as a build dependency, it should get linked in statically
```
Comment (boegel) [inline easybuild/easyblocks/b/binutils.py]: `should get linked` -> `should have been linked`

File: easybuild/easyblocks/b/binutils.py [hunk 2]
```diff
@@ -78,10 +78,15 @@ def configure_step(self):
...
+        # configure using `--with-system-zlib` if zlib is a (build) dependency
         zlibroot = get_software_root('zlib')
         if zlibroot:
             self.cfg.update('configopts', '--with-system-zlib')
+
+        # statically link to zlib only if it is a build dependency
+        # see https://github.com/easybuilders/easybuild-easyblocks/issues/1350
+        build_deps = self.cfg.dependencies(build_only=True)
+        if any(dep['name'] == 'zlib' for dep in build_deps):
             libz_path = os.path.join(zlibroot, get_software_libdir('zlib'), 'libz.a')
```
Comment (boegel) [inline easybuild/easyblocks/b/binutils.py]: @geimer This should be intended one level more, there's no point in checking whether `zlib` was a build-only dep if `zlibroot` is not defined (since `get_software_libdir` will crash hard then)

**Comment by boegel** [comment]
@geimer Since we're currently only including `zlib` as a build dep, this doesn't actually change much for existing easyconfigs (except when people have their own `binutils` easyconfigs where `zlib` is a runtime dep already), right?

I'm aware that we're changing that in https://github.com/easybuilders/easybuild-easyconfigs/pull/8340, just trying to make sure I have a good view on this w.r.t. backwards compatibility, since the idea is to include this with EasyBuild v3.9.1...

How much of a problem is it that `libz.a` is still embedded in `libbfd.a`?
In other words, if we merge this together with https://github.com/easybuilders/easybuild-easyconfigs/pull/8340, is #1350 actually fixed, or not?

**Comment by geimer** [comment]
> Since we're currently only including zlib as a build dep, this doesn't actually change much for existing easyconfigs (except when people have their own binutils easyconfigs where zlib is a runtime dep already), right?

It should be a drop-in replacement with existing easyconfigs, correct.

> How much of a problem is it that libz.a is still embedded in libbfd.a?

With the modified easyconfigs for newer `binutils`, this should only happen for `binutils` built with `dummy`, which is only used for bootstrapping `GCCcore` and `binutils` built with that `GCCcore` (and stuff explicitly using the `dummy`-built `binutils`, but these should be rare, no?).  Thus it's unlikely to be a huge problem.  If you want that also fixed, I currently only see two options: (a) require `zlib` as an OS dependency for the `dummy`-built `binutils` and not embed anything (thus dancing around the conflicts and `module switch` problems) or (b) mess around in the Makefiles even more. While (b) *may* be doable, it requires much more investigation, i.e., postponing the fix to some release in the future.

> In other words, if we merge this together with easybuilders/easybuild-easyconfigs#8340, is #1350 actually fixed, or not?

With both PRs merged, #1350 *should* be fixed for everything built on top of `GCCcore` using the `binutils` built with that `GCCcore`, as the latter no longer embeds `zlib`.

File: easybuild/easyblocks/b/binutils.py (hunk 2 above)
Comment (geimer) [reply easybuild/easyblocks/b/binutils.py]: It won't crash, as there isn't a way to list `zlib` as a build dependency without `zlibroot` being defined. But I agree, it could be indented.

**Comment by boegel** [comment]
Tested with a range of existing `binutils` easyconfigs old & new (which still have `zlib` as a build-only dep), no problems found.
Working test reports for https://github.com/easybuilders/easybuild-easyconfigs/pull/8340 are under way...

**Comment by boegel** [comment]
No surprises during testing, so going in, thanks @geimer!
